### PR TITLE
Update Data Generator

### LINKED ISSUE #3228: Update Data Generator to Use Desired Output Format
Based on our internal discussions, we want to modify the current output format of the data generator to include a JSON file with offsets and train/test/validation splits, as well as additional output for the class labels for those splits.

### PR DESCRIPTION
**Pull request by alexbarghi-nv**
Updates the generator to output offsets, labels, and train/test/validation splits.

Partially addresses https://github.com/rapidsai/graph_dl/issues/119

Closes #3228

### CODE REVIEW

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py [hunk 1]
```diff
@@ -0,0 +1,793 @@
...
+def setup_sg() -> None:
+
+    # Set RMM to allocate all memory as managed memory
+    rmm.mr.set_current_device_resource(rmm.mr.ManagedMemoryResource())
+    assert rmm.is_initialized()
+
+
+def setup_mg() -> None:
+    print("Setup MG")
+    client, cluster = start_dask_client()
```
Comment (VibhuJawa) [inline benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:139]: This causes OOM as you dont managed_memory here.

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py [hunk 2]
```diff
@@ -0,0 +1,793 @@
...
+    # --- Step 4:Author to Papers ---
+    num_auth = create_author_papers_data(args, (last_paper_id + 1), num_papers)
+
+    # --- Step 5:Author to Institutions ---
+    next_id = num_papers + num_auth + 1
+    num_inst = create_author_institute_data(args, (next_id), num_auth)
+
+    if args.mg:
+        print("Stopping cluster")
+        stop_mg(client, cluster)
```
Comment (VibhuJawa) [inline benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:765]: Move this down as the cluster tear-down can have bugs.

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py [hunk 3]
```diff
@@ -0,0 +1,826 @@
...
+                    int(last_paper_id + num_auth) + 1,
+                    int(last_paper_id + num_auth + num_inst) - 1
+                ],
+                'author__writes__paper': num_author_writes_paper_edges,
+                'author__affiliated_with__institution':
+                    num_author_affil_institution_edges,
+                'paper__cites__paper': num_paper_cites_paper_edges,
+                'train': train_ix.tolist(),
+                'test': test_ix.tolist(),
+                'val': val_ix.tolist(),
```
Comment (VibhuJawa) [inline benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:813]: We are writing these to disk as numpy arrays . I dont think we should write these in json as this wont scale and makes parsing a dict for large graphs prohibitive.

Comment (alexbarghi-nv) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:813]: We currently are not writing these to disk.  I would have to change that.

Comment (alexbarghi-nv) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:813]: The numpy files contain the train/test/val labels but not the indices.  But I can modify those files.

Comment (VibhuJawa) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:813]: > 1. We currently are not writing these to disk. I would have to change that.

I think we are , see lines:  https://github.com/rapidsai/cugraph/blob/6902ffea35b56b33b96bd3b12a1c27b6c6b23ebb/benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py#L792-L795


> The numpy files contain the train/test/val labels but not the indices. But I can modify those files.

Yes please. I wrote my sampler on the assumption that the train/test/ival  numpy files contains indices and not labels . I will modify it but we should have 2 files, one with **labels** and one with corresponding  **indices**  . 

We need to consistent mapping.

Comment (alexbarghi-nv) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:813]: I'm almost done fixing it, just give me a minute.  Why don't you want a single parquet with values and indices together?

Comment (VibhuJawa) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:813]: That also works.

Comment (alexbarghi-nv) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:813]: ok, great

Comment (alexbarghi-nv) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:813]: give me about a minute

Comment (alexbarghi-nv) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:813]: ok, should be good to go noe

Comment (VibhuJawa) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:813]: I see train_ix.tolist() , am i missing something ?

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py [hunk 4]
```diff
@@ -0,0 +1,822 @@
...
+        edge/
+            part.0.parquet
+            part.1.parquet
+            ...
+    paper/
+        node_feat.npy
+        node_label
+        test_labels
+        train_labels
+        val_labels
```
Comment (VibhuJawa) [inline benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:60]:
```suggestion
        node_label.parquet
        test_labels.parquet
        train_labels.parquet
        val_labels.parquet
```

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py [hunk 5]
```diff
@@ -0,0 +1,822 @@
...
+
+    meta.json
+
+The edge parquet files contain the src and dst arrays.
+
+node_feat.npy contains the node features
+and is of shape (# papers, num_features)
+
+node_label (parquet) is primarily for debugging.  It is
+a list of the node ids that are labeled.
```
Comment (VibhuJawa) [inline benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:70]: Is this true ?

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py [hunk 6]
```diff
@@ -0,0 +1,822 @@
...
+    nodes["label"].loc[nodes["rand"] < num_labeled] = 1
+
+    # selected just the labeld nodes
+    labeled_df = (nodes["label"].loc[nodes["label"] == 1]).to_frame().reset_index()
+
+    # drop the "labeled" column since it is no longer needed
+    labeled_df.drop(columns=["label"], inplace=True)
+
+    # Save the list
+    name = "node_label"
```
Comment (VibhuJawa) [inline benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:372]:
```suggestion
    name = "node_label.parquet"
```

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py [hunk 7]
```diff
@@ -0,0 +1,822 @@
...
+
+    test_labels = rng.integers(0, args.numClasses, len(test_ix))
+    test_df = pd.DataFrame({'node': test_ix, 'label': test_labels})
+
+    val_labels = rng.integers(0, args.numClasses, len(val_ix))
+    val_df = pd.DataFrame({'node': val_ix, 'label': val_labels})
+
+    save_data(args, paper_dir, 'train_labels', train_df)
+    save_data(args, paper_dir, 'test_labels', test_df)
+    save_data(args, paper_dir, 'val_labels', val_df)
```
Comment (VibhuJawa) [inline benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:439]:
```suggestion
    save_data(args, paper_dir, 'train_labels.parquet', train_df)
    save_data(args, paper_dir, 'test_labels.parquet', test_df)
    save_data(args, paper_dir, 'val_labels.parquet', val_df)
```

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Have requested that you provide extension to the files we write.

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py (hunk 5 above)
Comment (alexbarghi-nv) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:70]: It should be.  Is it not?

**Comment by alexbarghi-nv** [comment]
Based on internal discussion these files will not be committed to the cugraph repository.

**Pull request closed without merging**